src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs

```diff
@@ -474,12 +484,27 @@ public class OutputCacheFilter : FilterProvider, IActionFilter, IResultFilter, I
                 response.AddHeader("X-Cached-On", cacheItem.CachedOnUtc.ToString("r"));
                 response.AddHeader("X-Cached-Until", cacheItem.ValidUntilUtc.ToString("r"));
             }
-
+            
             // Shorcut action execution.
             filterContext.Result = new FileContentResult(cacheItem.Output, cacheItem.ContentType);
-
             response.StatusCode = cacheItem.StatusCode;
 
+            // Add ETag header
+            if (HttpRuntime.UsingIntegratedPipeline && response.Headers.Get("ETag") == null) {
+                response.Headers["ETag"] = cacheItem.ETag;
+            }
+
+            // Check ETag in request
+            // https://www.w3.org/2005/MWI/BPWG/techs/CachingWithETag.html
+            var etag = request.Headers["If-None-Match"];
+            if (!String.IsNullOrEmpty(etag)) {
+                if (String.Equals(etag, cacheItem.ETag, StringComparison.Ordinal)) {
+                    // ETag matches the cached item, we return a 304
+                    filterContext.Result = new HttpStatusCodeResult(304, "Not Modified");
```

Comment:
"Not Modified" is the default for 304, so we should use the overload without the string.